File: simulator/abilities/on_hit_divine_smite.py
from ..abilities.on_hit_effect import OnHit
from ..misc import Class, avg_roll, _roll_dice, DamageType
import logging

logger = logging.getLogger("Encounterra")


class OnHitDivineSmite(OnHit):

    @staticmethod
    def get_dmg_dice(spell_slot_level):
        match spell_slot_level:
            case 1:
                return (2, 8)
            case 2:
                return (3, 8)
            case 3:
                return (4, 8)
            case 4:
                return (5, 8)
            case _:
                logger.error("Incorrect Divine Smite Level")
                return (2, 8)

    @staticmethod
    def get_dmg_dice_undead_or_fiend(spell_slot_level):
        match spell_slot_level:
            case 1:
                return (3, 8)
            case 2:
                return (4, 8)
            case 3:
                return (5, 8)
            case 4:
                return (6, 8)
            case _:
                logger.error("Incorrect Divine Smite Level")
                return (3, 8)

    def __init__(self, name="Divine Smite"):
        self.name = name

    def hit(self, attacker, attack, target, multiplier, dmg_so_far):
        for level in range(4, 0, -1):
            if attacker.spellslots.has_resource(level=level):
                target_hp = target.curr_hp
                dmg_dice = OnHitDivineSmite.get_dmg_dice_undead_or_fiend(level) if (type(target).cls is Class.MONSTER.UNDEAD or type(target).cls is Class.MONSTER.FIEND) else OnHitDivineSmite.get_dmg_dice(level)
                avg_dmg = avg_roll(dmg_dice)
                if (target_hp - dmg_so_far) * 1.3 >= avg_dmg * multiplier:
                    attacker.spellslots.use_resource(level=level)
                    logger.error(f"{attacker} uses Divine Smite of level {level} on {target}")
                    return [_roll_dice(dmg_dice) * multiplier, DamageType.Radiant]
        return None

    def calculate_threat(self, attacker, target, **kwargs):
        dmg_acc = 0
        count = 0
        for level in range(4, 0, -1):
            if attacker.spellslots.has_resource(level=level):
                dmg_dice = OnHitDivineSmite.get_dmg_dice_undead_or_fiend(level) if (type(target).cls is Class.MONSTER.UNDEAD or type(target).cls is Class.MONSTER.FIEND) else OnHitDivineSmite.get_dmg_dice(level)
                dmg_acc = avg_roll(dmg_dice)
                count += 1
        if count:
            dmg_acc /= count
        return dmg_acc

File: c++/abilities/on_hit_divine_smite.cpp
#include "abilities/on_hit_divine_smite.hpp"

#include "core/battle_map.hpp"
#include "core/combatant.hpp"
#include <algorithm>
#include <iostream>

namespace enc
{
  Die OnHitDivineSmite::getDmgDice(int spellSlotLevel)
  {
    return Die{static_cast<unsigned char>(std::clamp(spellSlotLevel + 1, 2, 5)), 8};
  }

  Die OnHitDivineSmite::getDmgDiceUndeadOrFiend(int spellSlotLevel)
  {
    return Die{static_cast<unsigned char>(std::clamp(spellSlotLevel + 2, 3, 6)), 8};
  }

  bool OnHitDivineSmite::isUndeadOrFiend(Combatant *target)
  {
    Combatant *form = target->getOriginalForm();
    return form->isMonsterType(Monster::UNDEAD) || form->isMonsterType(Monster::FIEND);
  }

  bool OnHitDivineSmite::canSmite(Combatant *attacker)
  {
    if(attacker == nullptr || !attacker->hasPassiveAbility(AbilityType::DIVINE_SMITE))
      {
        return false;
      }
    if(auto freeSmite = attacker->getResource(AbilityType::DIVINE_SMITE); freeSmite && (*freeSmite)->hasUses())
      {
        return true;
      }
    if(attacker->hasAlreadyUsedSpellslotThisTurn())
      {
        return false;
      }
    for(int level = 4; level >= 1; --level)
      {
        if(attacker->getSpellslots().hasUses(level))
          {
            return true;
          }
      }
    return false;
  }

  int OnHitDivineSmite::chooseSmiteLevel(Combatant *attacker, Combatant *target, double multiplier, double dmgSoFar)
  {
    if(!canSmite(attacker))
      {
        return 0;
      }
    if(auto freeSmite = attacker->getResource(AbilityType::DIVINE_SMITE); freeSmite && (*freeSmite)->hasUses())
      {
        return 1;
      }

    for(int level = 4; level >= 1; --level)
      {
        if(attacker->getSpellslots().hasUses(level))
          {
            Die dice = isUndeadOrFiend(target) ? getDmgDiceUndeadOrFiend(level) : getDmgDice(level);
            double avgDmg = avgRoll(dice);
            if((target->getCurrentHp() - dmgSoFar) * 1.3 >= avgDmg * multiplier)
              {
                return level;
              }
          }
      }
    return 0;
  }

  std::vector<std::pair<int, DamageType>>
  OnHitDivineSmite::hit(Combatant *attacker, Actoid * /*attack*/, Combatant *target, double multiplier, double dmgSoFar)
  {
    if(attacker == nullptr || target == nullptr)
      {
        return {};
      }

    int chosenLevel = chooseSmiteLevel(attacker, target, multiplier, dmgSoFar);
    if(chosenLevel == 0)
      {
        return {};
      }

    if(auto freeSmite = attacker->getResource(AbilityType::DIVINE_SMITE); freeSmite && (*freeSmite)->hasUses())
      {
        (*freeSmite)->useResource();
      }
    else
      {
        attacker->getSpellslots().useResource(chosenLevel);
        attacker->setAlreadyUsedSpellslotThisTurn(true);
      }

    Die dice = isUndeadOrFiend(target) ? getDmgDiceUndeadOrFiend(chosenLevel) : getDmgDice(chosenLevel);
    int dmg = rollDice(dice);
    if(multiplier >= 2)
      {
        dmg *= 2;
      }
    std::cout << attacker->_name << " uses Divine Smite of level " << chosenLevel << " on " << target->_name << " for " << dmg
              << " Radiant damage" << std::endl;
    return {{dmg, DamageType::Radiant}};
  }

  double OnHitDivineSmite::calculateThreat(Combatant *attacker, Combatant *target, RollType rollType)
  {
    int chosenLevel = chooseSmiteLevel(attacker, target);
    if(chosenLevel == 0)
      {
        return 0.0;
      }
    return avgRoll(isUndeadOrFiend(target) ? getDmgDiceUndeadOrFiend(chosenLevel) : getDmgDice(chosenLevel));
  }
}
